### PR TITLE
Allowance budget mutated

### PR DESCRIPTION
**Pull request by lunfardo314**
Now `Allowance` budget in the `Sandbox` is mutated mutated with each call to `sandbox.TransferAllowedFunds`. 
It is needed to keep track of the allowed budget and prevent overrun by repeating calls to `TransferAllowedFunds`

### CODE REVIEW

File: packages/iscp/assets.go [hunk 1]
```diff
@@ -200,6 +200,7 @@ func (a *Assets) Equals(b *Assets) bool {
 // FitsTheBudget checks if:
 // - 'a' has all non-negative values. Negative values is error
 // - 'a' all values are <= of corresponding values of 'b'. It means, we can debit 'a' from 'b' without overrun of funds
+// Deprecated: not needed anymore
```
Comment (jorgemmsilva) [inline packages/iscp/assets.go:203]: can we remove the deprecated methods?

File: packages/vm/core/accounts/impl.go [hunk 2]
```diff
@@ -75,24 +75,29 @@ func withdraw(ctx iscp.Sandbox) (dict.Dict, error) {
...
 	if ctx.Caller().Address().Equal(ctx.ChainID().AsAddress()) {
 		// if the caller is on the same chain, do nothing
 		return nil, nil
 	}
 	// move all allowed funds to the account of the current contract context
+	// before saving the allowance budget because after the transfer it is mutated
+	fundsToWithdraw := ctx.AllowanceAvailable().Clone()
 	ctx.TransferAllowedFunds(ctx.AccountID())
 
+	// por las dudas
```
Comment (jorgemmsilva) [inline packages/vm/core/accounts/impl.go:89]: :sweat_smile: ?

**Review by jorgemmsilva — APPROVED** [review]

File: packages/vm/vmcontext/runreq.go [hunk 3]
```diff
@@ -194,7 +194,7 @@ func (vmctx *VMContext) calculateAffordableGasBudget() {
 	if vmctx.req.SenderAddress() == nil {
 		panic("inconsistency: vmctx.req.SenderAddress() == nil")
 	}
-	// calculate how many tokens for gas fee can be guaranteed after taking into account the allowance
+	// calculate how many tokens for gas fee can be guaranteed after taking into account the allowanceRemains
```
Comment (dessaya) [inline packages/vm/vmcontext/runreq.go:197]: unintended rename?

File: packages/vm/vmcontext/vmcontext.go [hunk 4]
```diff
@@ -69,10 +69,10 @@ type VMContext struct {
...
 
 type callContext struct {
-	caller    *iscp.AgentID // calling agent
-	contract  iscp.Hname    // called contract
-	params    dict.Dict     // params passed
-	allowance *iscp.Assets  // allowance passed
+	caller           *iscp.AgentID // calling agent
+	contract         iscp.Hname    // called contract
+	params           dict.Dict     // params passed
+	allowanceRemains *iscp.Assets  // MUTABLE: allowance budget left after TransferAllowedFunds
```
Comment (dessaya) [inline packages/vm/vmcontext/vmcontext.go:75]: `allowanceAvailable` better? To make it consistent with the `AllowanceAvailable()` method

File: packages/vm/vmcontext/call.go [hunk 5]
```diff
@@ -67,10 +67,10 @@ const traceStack = false
...
 func (vmctx *VMContext) pushCallContext(contract iscp.Hname, params dict.Dict, allowance *iscp.Assets) {
 	ctx := &callContext{
-		caller:    vmctx.getToBeCaller(),
-		contract:  contract,
-		params:    params.Clone(),
-		allowance: allowance,
+		caller:           vmctx.getToBeCaller(),
+		contract:         contract,
+		params:           params.Clone(),
+		allowanceRemains: allowance.Clone(), // we have to clone it because it will be mutated by TransferAllowedFunds
```
Comment (dessaya) [inline packages/vm/vmcontext/call.go:73]: How is a call between contracts handled?

I have an allowance A = 10
I call contract C1. The allowance is cloned, so I have A1 = 10
Contract C1 calls contract C2. The allowance is cloned, so I have A2 = 10
Contract C2 spends 5, so A2 = 5.
When the call context is popped, the current allowance for contract C1 is still A1 = 10?

File: packages/iscp/assets.go (hunk 1 above)
Comment (lunfardo314) [reply packages/iscp/assets.go:203]: I will, I promise

File: packages/vm/vmcontext/call.go (hunk 5 above)
Comment (lunfardo314) [reply packages/vm/vmcontext/call.go:73]: yes. Allowance is what C1 (caller) allows for C2 (target) to take from C1's account. Allowance is not passed, it is intended for the target. If target calls another target, it creates another, unrelated, allowance wrt its own account

**Pull request merged**